### PR TITLE
report status to stderr instead of stdout

### LINKED ISSUE #1473: Cargo output could go to stderr
Currently output generated by `cargo` itself goes to stdout, while `rustc`s goes to stderr. This is pretty annoying for a command `cargo run`, where the user controls stdout and may not appreciate the contamination from `cargo`, e.g. `cargo run --bin foo > some_file.txt` will include the `Compiling ...` and `Running ...` lines.

Alteratively, I suppose we could support `cargo run --output some_file.txt --bin foo` which will do the redirect internally.

### PR DESCRIPTION
**Pull request by matklad**
All status messages are printed to stderr instead of stdout now.

This changes exactly three letters of code and breaks 211 tests :fearful: 

I will hopefully fix all the tests here, but I decided to submit PR early in case somebody comes up with a less laborious way to do this. 

Fixes #1473, #2661

### CODE REVIEW

**Comment by matklad** [comment]
> breaks 211 tests 

... on the stable compiler. There should be even for failures for nightly

**Comment by matklad** [comment]
91 failures left :)

**Comment by matklad** [comment]
Hm, I found an interesting phenomena: some tests before this changed asserted `stdout` and ignored `stderr` although  some warnings were printed to stderr. When converting `with_stdout` to `with_stderr` this test fail because there is more output. One way to fix them is to add warnings to the expected output, like I did in [this commit]. But is this really the best way? Could changing `with_stderr` to `with_stderr_contains` be a better fix? I think it would, because it makes tests more robust. 

[this commit]: https://github.com/rust-lang/cargo/pull/2693/commits/5574ca2f479f8f92f4f625b79c4df1a90f648382

**Comment by alexcrichton** [comment]
@matklad yeah that's probably a split between "intentionally ignored", "too lazy to write out the output", or "oops forgot to assert that output". It should be fine to either add the assertions or switch to `with_stdout_contains`.

Also cc @rust-lang/tools, a somewhat large-ish semantic change to Cargo!

**Comment by matklad** [comment]
>Also cc @rust-lang/tools, a somewhat large-ish semantic change to Cargo!

Nice, I indeed feel a little uneasy because of the scope of the change (Although I personally think that stderr is the right stream for the status messages). I'll then wait for an explicit approval before I continue fixing the tests :)

**Comment by nrc** [comment]
+1 for more disciplined use of stderr/out

**Comment by brson** [comment]
I agree that stderr is the right place for status messages. How will this affect `cargo run -q` (which I use because cargo prints to stdout)? I guess it will continue to not print things to stderr? Is that still useful?

**Comment by alexcrichton** [comment]
@brson in theory it won't affect `cargo run -q` as Cargo will continue to not print anything.

**Comment by alexcrichton** [comment]
Discussed during tools triage we were all on board, so let's do this! Wanna fix up the few remaining tests @matklad and I'll merge?

**Comment by matklad** [comment]
>Wanna fix up the few remaining tests @matklad and I'll merge?

Yeah, I'll do it, but I can't give a precise ETA (this weak perheps?). The remaining tests most likely all require some manual intervention.

**Comment by alexcrichton** [comment]
Ah yeah no worries, and if you run out of time I can help out as well

File: tests/test_cargo_compile_path_deps.rs [hunk 1]
```diff
@@ -722,7 +722,7 @@ test!(dev_deps_no_rebuild_lib {
         "#)
         .file("src/lib.rs", r#"
             #[cfg(test)] extern crate bar;
-            #[cfg(not(test))] fn foo() { env!("FOO"); }
+            #[cfg(not(test))] pub fn foo() { env!("FOO"); }
```
Comment (matklad) [inline tests/test_cargo_compile_path_deps.rs]: @alexcrichton this tests looks a bit strange. Looks like it is supposed to set and read an environmental variable, but it actually is never asserted (and "unused function foo" is printed during the compilation).

Perhaps the test doesn't test what it is supposed to?

**Comment by matklad** [comment]
@alexcrichton I think I have fixed all the tests except the cross compile ones (one does not simply link stuff on nixos :) ). Can you please finish the work here? I think the commits should be squashed before merge, but it maybe simpler to review them one by one.

File: tests/test_cargo_compile_path_deps.rs (hunk 1 above)
Comment (alexcrichton) [inline tests/test_cargo_compile_path_deps.rs]: I believe the `env!` macro fails compilation if the env var isn't defined

**Pull request merged**